### ISSUE TITLE
How to change the code to train my own dataset?

### ISSUE THREAD

**Issue by zhoulukuan** [Superlee506/Mask_RCNN#1]
Thanks for your great work. In fact, I want to use mask rcnn to predict keypoints in clothes which have more classes and keypoints. I have two questions:
1. Due to the different formats of train/test data, I must change it to the format of COCO annotations, right?
2. More classes and keypoints lead to different networks and the ways of processing. Do I need to change the code of network/loss or other places? Can you give me some help?
Thank you in advance.

**Comment by Superlee506** [comment]
Mask RCNN for human pose estimation just has 2 classes(1 for background 1 for person). So as to your question. You just need to change  the following functions:
1. In fpn_classifier_graph, you just need to change the num_classes
2. In build_fpn_mask_graph, you also just need to change the num_classes
3. In build_fpn_keypoint_graph, because I use it to predict the keypoint of human, you have to change the output as [batch_size, num_roi, num_classes, num_keypoints].
4. In keypoint_mrcnn_mask_loss_graph, you need to add some code to get the class specific mask of each ROI as mrcnn_mask_loss_graph does.

**Comment by Superlee506** [comment]
@zhoulukuan

**Comment by zhoulukuan** [comment]
@Superlee506 Thanks a lot. I will have a try soon.

**Comment by zhoulukuan** [comment]
@Superlee506  I see that there is still mask loss in your code. Do you use segmentation label to improve result? If I only have classes label and keypoints label and don't have segmentation label, what should I do?

I don't konw if it is a good idea to remove the mask_loss in the following places directly.
```python
mask_loss = KL.Lambda(lambda x: mrcnn_mask_loss_graph(*x),
                                           name="mrcnn_mask_loss")(
                [target_mask, target_class_ids, mrcnn_mask])
keypoint_loss = KL.Lambda(lambda x: keypoint_mrcnn_mask_loss_graph(*x, weight_loss=config.WEIGHT_LOSS), name="keypoint_mrcnn_mask_loss")(
                [target_keypoint, target_keypoint_weight, target_class_ids, keypoint_mrcnn_mask])
```

```python
# add "test_target_keypoint_mask" in the output for test the keypoint loss function
            outputs = [rpn_class_logits, rpn_class, rpn_bbox,
                       mrcnn_class_logits, mrcnn_class, mrcnn_bbox, keypoint_mrcnn_mask,
                       rpn_rois, output_rois,
                       rpn_class_loss, rpn_bbox_loss, class_loss, bbox_loss, keypoint_loss, mask_loss]
                       # +  test_target_keypoint_mask for test the keypoint loss graph

            model = KM.Model(inputs, outputs, name='mask_keypoint_mrcnn')
```

**Comment by zhoulukuan** [comment]
@Superlee506 Besides, I have some questions about "keypoint_mrcnn_mask_loss_graph":
1. What's the meaning of variable target_keypoint_weights? 
2. According to code annotation, you convert the target mask to a number which represent positions in the mask. Dose it cause some difference?

**Comment by Superlee506** [comment]
@zhoulukuan 
1. You can remove the mask head directly, it dosen't matter and will have very little influence on the final prediction (If all the other things are correct).
2. The keypoint_weights is used to indicate wheter the keypoint is contained or visible in the ROI.
3. I use a label instead of the keypoint mask, because keypoint mask will have more than 1 values after cropping and resizing. The detecron source code also does the same thing.

**Comment by zhoulukuan** [comment]
Thank you very much. When I try to change the code, I meet a question about my dataset. In fact, different class has different key points, I don't have a clear idea on how to deal with it. Should I consider class label and only compute loss of those key points which this kind has?
According to your code, I guess positive_keypoint_weights can be used to remove impact of those unvisible keypoints. If so, can I set those inexistent keypoints's weight to zero for removing impact？

**Comment by Superlee506** [comment]
You can set the largest keypoint number as the output and use keypoint_weight to ignore the invisible points.

**Comment by Superlee506** [comment]
@zhoulukuan

**Comment by zhoulukuan** [comment]
Every class should be have its own 24 keypoints instead of  sharing 24 keypoints together for decouple?

**Comment by zhoulukuan** [comment]
Every class should have its own 24 keypoints instead of sharing 24 keypoints together for decouple?

**Comment by JJJackssson** [comment]
> Mask RCNN for human pose estimation just has 2 classes(1 for background 1 for person). So as to your question. You just need to change the following functions:
> 
> 1. In fpn_classifier_graph, you just need to change the num_classes
> 2. In build_fpn_mask_graph, you also just need to change the num_classes
> 3. In build_fpn_keypoint_graph, because I use it to predict the keypoint of human, you have to change the output as [batch_size, num_roi, num_classes, num_keypoints].
> 4. In keypoint_mrcnn_mask_loss_graph, you need to add some code to get the class specific mask of each ROI as mrcnn_mask_loss_graph does.

In build_fpn_keypoint_graph, the original output shape is [batch, roi_count, num_keypoints, height*width], I need to change it as [batch_size, num_roi, num_classes, num_keypoints] or [batch_size, num_roi, num_classes, num_keypoints, height*width ] , are there any other functions' output needed to change?
